_posts/2020-09-30-sparklyr-1.4.0-released/sparklyr-1.4.0.Rmd

```diff
@@ -0,0 +1,368 @@
+
+An alternative way of standardizing $X$ based on its median, 1st quartile, and 3rd quartile values, all of which are robust against outliers, would be the following:
+
+$\displaystyle z = \frac{X - \text{Median}(X)}{\text{P75}(X) - \text{P25}(X)}$
+
+and this is precisely what [RobustScaler](https://spark.apache.org/docs/3.0.0/api/java/org/apache/spark/ml/feature/RobustScaler.html) offers.
+
+To see `ft_robust_scaler()` in action and demonstrate its usefulness, we can go through a contrived example consisting of the following steps:
+
+- Draw 100 random samples from the standard normal distribution
+
+```{r echo=TRUE, eval=FALSE}
+sample_values <- rnorm(100)
+print(sample_values)
+```
+
+```
+  [1] -0.885440337  0.831855863 -0.890427366  0.437780168  0.091902152
+  [6]  0.774895164 -0.839350781 -1.173025189 -0.442174680 -0.002573251
+  ...
+```
+- Inspect the minimal and maximal values among the 100 random samples:
+
+```{r echo=TRUE, eval=FALSE}
+print(min(sample_values))
+```
+```
+  [1] -2.377851
+```
+
+```{r echo=TRUE, eval=FALSE}
+print(max(sample_values))
+```
+```
+  [1] 2.629998
+```
+
+- Now create 5 other values that are extreme outliers compared to the 100 random samples above. Given that we know all 100 samples are within the range of $(-3, 3)$, we can choose $21, 22, \ldots, 25$ as our outliers:
+
+```{r echo=TRUE, eval=FALSE}
+outliers <- 20 + seq(5)
+```
+
+- Copy all 105 values into a Spark dataframe named `sdf`
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "3.0.0")
+sdf <- copy_to(sc, data.frame(value = c(sample_values, outliers)))
+```
+
+- We can then apply `ft_robust_scaler()` to obtain the standardized value for each input:
+
+```{r echo=TRUE, eval=FALSE}
+scaled <- sdf %>%
+  ft_vector_assembler("value", "input") %>%
+  ft_robust_scaler("input", "scaled") %>%
+  dplyr::pull(scaled) %>%
+  unlist()
+```
+
+- Plotting the result shows the non-outlier data points are scaled to values that still more or less form a bell-shaped distribution, as expected, so the scaling is robust against influences of the outliers:
+
+```{r echo=TRUE, eval=FALSE}
+library(ggplot2)
+
+ggplot(data.frame(scaled = scaled), aes(x = scaled)) + geom_histogram(binwidth = 0.2)
+```
+
+![](images/scaled.png){#id .class width=40% height=40%}
+
+- Finally, we can compare the distribution of the scaled values above with the distribution of z-scores of all input values, and notice how scaling the input with only mean and standard deviation would have caused noticeable skewness which the robust scaler has successfully avoided:
+
+
+```{r echo=TRUE, eval=FALSE}
+all_values <- c(sample_values, outliers)
+z_scores <- (all_values - mean(all_values)) / sd(all_values)
+ggplot(data.frame(scaled = z_scores), aes(x = scaled)) + geom_histogram(binwidth = 0.2)
+```
+
+![](images/skewed.png){#id .class width=40% height=40%}
```

Comment:
Same for this plot as the comment before, I would make it wider to use the full paragraph size and change it to a lighter theme, but not really that important.

Reply:
Good idea!

Re: making plots wider -- I don't think it's strictly needed. Both distributions will be bell-shaped, but the latter will not be centered around 0. What I did instead was changing the wording in this section to emphasis that fact more, and also changing the `binwidth` in the second plot to reflect that visually.